README.md
# act-sql-common

[![APL v2](https://img.shields.io/badge/license-Apache%202-blue.svg)](http://www.apache.org/licenses/LICENSE-2.0.html) 
[![Maven Central](https://img.shields.io/maven-central/v/org.actframework/act-sql-common.svg)](http://search.maven.org/#search%7Cga%7C1%7Ca%3A%22act-sql-common%22)
[![Build Status](https://travis-ci.org/actframework/act-sql-common.svg?branch=master)](https://travis-ci.org/actframework/act-sql-common)
[![codecov](https://codecov.io/gh/actframework/act-sql-common/branch/master/graph/badge.svg)](https://codecov.io/gh/actframework/act-sql-common)
[![Javadocs](http://www.javadoc.io/badge/org.actframework/act-sql-common.svg?color=blue)](http://www.javadoc.io/doc/org.actframework/act-sql-common)

Common module for SQL database plugin

## Versions

| *Actframework version* | *act-sql-common version* |
| ---------------------- | ------------------------ |
| 1.1.0 | 1.0.0 |
| 1.1.1 | 1.0.1 |

## Configurations

### 1. Data source configuration

#### `url` or `jdbcUrl` or `databaseUrl`

Specifies the jdbc url of the datasource. 

Default value: `jdbc:h2:./test` if h2 jdbc driver is provided, otherwise it will break the application from bootstrap.

#### `driver` or `jdbcDriver` or `databaseDriver`

Optionally. Specifies the jdbc driver class. If not specified the framework will infer the driver class from the `url`

#### `username`

Specifies the jdbc connection username

Default value: `sa`

#### `password`

Specifies the jdbc connection password

Default value: empty string

#### `datasource` or `datasource.provider` 

Specifies the `DataSourceProvider` implementation.

About value and default value: 

1. if `hikari` found in the value then it will use the `act.db.sql.datasource.HikariDataSourceProvider`
1. otherwise if `druid` found in the value then it will use the `act.db.sql.datasource.DruidDataSourceProvider`
1. otherwise if the value starts with `shared:` then it will try to extract the `dbId` from the value, e.g. if value is `shared:ds1` then the `dbId` is `ds1`, after that it will try to load the db service with id `ds1` and use the datasource of that db service and feed into the current db service being configured. This allows multiple db service share the same datasource instance. **Note** app must make sure the db service been referenced has already configured and the db service must be an implementation of `SqlDbService`. If a datasource is configured as shared datasource then it does not need to configure other datasource properties e.g. url, username etc. However the `db.naming.convention` still needs to be configured to make the db service able to mapping between table/column and class/field.
1. otherwise if specified then it will try to load the `DataSourceProvider` instance by using `Act.getInstance(<value>)`
1. if not specified then system will probe the class loader by try loading `com.zaxxer.hikari.HikariConfig` or `com.alibaba.druid.pool.DruidDataSource` and load the hikari CP provider or druid provider respectively

#### `autoCommit`

Specify should datasource auto commit transaction

Default value: `false`

#### `minConnections`

Specify the minimum connections should be created by data source

Default value: `2`
 
#### `maxConnections`

Specify the maximum connections can be created by data source

Default value: `10`

#### `isolationLevel`

Specifies the default transaction isolation level, could be one of the following:

1. `NONE` - no transaction
1. `READ_UNCOMMITTED`
1. `READ_COMMITTED`
1. `REPEATABLE_READ`
1. `SERIALIZABLE`

Default value: `READ_COMMITTED`

####  `db.naming.convention`

Specify the naming convention to map Entity class name to table name and the property name to column name
 
Supported values:

1. `matching` - map directly (could be case insensitive)
1. `underscore` - convert camelcase to underscore notation

Default value: `matching`

### 2. DDL configuration

The ddl configuration tells the framework whether or not to generate and run DDL scripts

#### `ddl.create`

Specify should framework to generate and run create table DDL
 
Default value: `false` unless `h2` is used as database

**Note** about `h2`: if using h2 anbd the database file exists then `ddl.create` will always be treated as `false` unless `ddl.drop` is specified as `true`

#### `ddl.drop`

Specify should framework to generate and run drop table DDL
 
Default value: `false`

### 3. HikariCP specific configuration

#### 3.1 `idleTimeout`

```java
String s = miscConf.get("idleTimeout");
if (null != s) {
    int n = Integer.parseInt(s);
    hc.setIdleTimeout(n);
} else {
    hc.setIdleTimeout(conf.maxInactiveTimeSecs * 1000);
}
```

#### 3.2 `connectionInitSql`

```java
s = miscConf.get("connectionInitSql");
if (null != s) {
    hc.setConnectionInitSql(s);
}
```

#### 3.3 `maxLifetime`

```java
s = miscConf.get("maxLifetime");
if (null != s) {
    long n = Long.parseLong(s);
    hc.setMaxLifetime(n);
} else {
    hc.setMaxLifetime(conf.maxAgeMinutes * 60 * 1000L);
}
```

#### 3.4 `poolName`

```java
s = miscConf.get("poolName");
if (null != s) {
    hc.setPoolName(s);
}
```

### 4. Druid specific configuration

#### 4.1 `initialSize`

```java
DruidDataSource source = new DruidDataSource();
String s = miscConf.get("initialSize");
if (null != s) {
    source.setInitialSize(Integer.parseInt(s));
} else {
    source.setInitialSize(source.getMinIdle());
}
```

#### 4.2 `timeBetweenEvictionRunsMillis`

```java
s = miscConf.get("timeBetweenEvictionRunsMillis");
if (null != s) {
    source.setTimeBetweenEvictionRunsMillis(Long.parseLong(s));
}
```

#### 4.3 `minEvictableIdleTimeMillis`

```java
s = miscConf.get("minEvictableIdleTimeMillis");
if (null != s) {
    source.setMinEvictableIdleTimeMillis(Long.parseLong(s));
}
```

#### 4.4 `testWhileIdle`

```java
s = miscConf.get("testWhileIdle");
if (null != s) {
    source.setTestWhileIdle(Boolean.parseBoolean(s));
}
```

#### 4.5 `testOnBorrow`

```java
s = miscConf.get("testOnBorrow");
if (null != s) {
    source.setTestOnBorrow(Boolean.parseBoolean(s));
}
```

#### 4.6 `testOnReturn`

```java
s = miscConf.get("testOnReturn");
if (null != s) {
    source.setTestOnReturn(Boolean.parseBoolean(s));
}
```

#### 4.7 `filters`

```java
s = miscConf.get("filters");
if (null != s) {
    try {
        source.setFilters(s);
    } catch (SQLException e) {
        throw E.unexpected(e);
    }
}
```

#### 4.8 `poolPreparedStatements`

```java
s = miscConf.get("poolPreparedStatements");
if (null != s) {
    source.setPoolPreparedStatements(Boolean.parseBoolean(s));
}
```


CHANGELOG.md
# act-sql-common CHANGE LOG

1.6.1
* Default URL conflict with multiple data sources #28

1.6.0 - 03/Nov/2019
* update to act-1.8.29

1.5.1 - 30/Sep/2018
* update to act-1.8.28
* Tune jobId for all jobs

1.5.0 - 2/Jul/2018
* update to act-1.8.25
* Allow inject SqlDbService and DataSource #23

1.4.6 - 16/Jun/2019
* Error starting with db.default.datasource.provider configuration #22

1.4.5 - 20/Apr/2019
* update to act-1.8.20

1.4.4
* update to act-1.8.9
* Update SqlDbService init logic #18
* update SqlDbService for
* Make TX follow JPA `TxType.REQUIRED` semantic #17
* Apply new scope for implicit transaction #16

1.4.3 - 30/Oct/2018
* update to act-1.8.8

1.4.2 - 19/Jun/2018
* `TxInfo` - exitTxScope logic issue #15
* update to act-1.8.8-RC10
* update druid to 1.1.10

1.4.1 - 7/Jun/2018

* update to act-1.8.8-RC9
* `SqlDbPlugin.applyTo(App)` issue #14

1.4.0 - 29/May/2018

* It shall ignore the Exception raised by data source during shutdown process. #12

1.3.4 - 15/May/2018
* Support auto setup driver for MariaDB #10
* update act to 1.8.8-RC5

1.3.3 - 02/Apr/2018
* update act to 1.8.5

1.3.2 - 25/Mar/2018
* Update act to 1.8.2
* Update druid to 1.1.9

1.3.1 - 11/Mar/2018
* merge from 1.2.1 for `HikariCP.setConnectionInitSql`

1.3.0
* Update to act-1.7.x
* Update to JPA 2.2
* Add JTA into dependency

1.2.1
* Support `HikariCP.setConnectionInitSql` configuration #8

1.2.0
* update act to 1.6.0

1.1.3
* update act to 1.5.1
* update druid to 1.1.5
* update HikariCP to 2.7.3

1.1.2
* catch up to act-1.4.11
* improve maven build process
* update hikaricp and druid version

1.1.1
* fix for ebean2 issue: NPE raised starting app in prod mode when no third part datasource defined #7

1.1.0
* catch up to act-1.4.0

1.0.2
- Allow specific implementation to initialize in different logic when dataSourceProvider available or not #3 

1.0.1
- The default jdbc driver doesn't work with mysql jdbc driver 5.x #2 
- When it uses h2 with db on filesystem, it shall ignore the `ddl.create` if that file exists #1 

1.0.0 - baseline version


src/main/java/act/db/sql/DataSourceConfig.java
package act.db.sql;

/*-
 * #%L
 * ACT AAA Plugin
 * %%
 * Copyright (C) 2015 - 2017 ActFramework
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import static act.Act.LOGGER;
import static act.db.sql.SqlConfKeys.*;

import act.app.DbServiceManager;
import act.data.annotation.Data;
import act.util.SimpleBean;
import org.osgl.$;
import org.osgl.exception.ConfigurationException;
import org.osgl.util.*;

import java.sql.Connection;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * The simple class encapsulate configurations for an SQL DataSource.
 *
 * All time configuration should be in seconds.
 */
@Data
public class DataSourceConfig implements SimpleBean {

    public static final int DEF_MIN_CONN = 2;
    public static final int DEF_MAX_CONN = 10;
    public static final int DEF_INACTIVE_TIME = 10 * 60;
    public static final int DEF_MAX_LIFE_TIME = 30 * 60;
    public static final int DEF_CONN_TIMEOUT = 30;

    public String id;

    public String url;

    public String username;

    public String password;

    public String driver;

    public int minConnections;

    public int maxConnections;

    public int isolationLevel = Connection.TRANSACTION_READ_COMMITTED;

    public boolean autoCommit;

    public boolean readOnly;

    public int connectionTimeout;

    public Map<String, String> customProperties;

    public List<DataSourceConfig> slaveDataSourceConfigurations = new ArrayList<>();

    // Constructor for slave datasource
    private DataSourceConfig(DataSourceConfig parent, Map<String, String> conf) {
        $.copy(parent).to(this);
        init(parent.id, conf, true);
        readOnly = true;
    }

    public DataSourceConfig(String dbId, Map<String, String> conf) {
        id = dbId;
        init(id, conf, false);

        // process for readonly data source
        Map<String, String> soleSlaveConfig = new HashMap<>();
        Map<String, Map<String, String>> slaves = new HashMap<>();
        for (Map.Entry<String, String> entry : conf.entrySet()) {
            String key = entry.getKey();
            String lowercaseKey = key.toLowerCase();
            if (lowercaseKey.startsWith("readonly.") || lowercaseKey.startsWith("ro.") || lowercaseKey.startsWith("slave")) {
                key = S.cut(key).afterFirst(".");
                String probe = S.cut(key).beforeFirst(".");
                if (probe.length() == 1) {
                    key = S.cut(key).afterFirst(".");
                    Map<String, String> slaveConfig = slaves.get(probe);
                    if (null == slaveConfig) {
                        slaveConfig = new HashMap<>();
                        slaves.put(probe, slaveConfig);
                    }
                    slaveConfig.put(key, entry.getValue());
                } else {
                    soleSlaveConfig.put(key, entry.getValue());
                }
            }
        }
        if (!soleSlaveConfig.isEmpty()) {
            slaveDataSourceConfigurations.add(new DataSourceConfig(this, soleSlaveConfig));
        }
        for (Map<String, String> otherSlaveConf : slaves.values()) {
            slaveDataSourceConfigurations.add(new DataSourceConfig(this, otherSlaveConf));
        }
    }

    public boolean hasSlave() {
        return !slaveDataSourceConfigurations.isEmpty();
    }

    public DataSourceConfig soleSlave() {
        return hasSoleSlave() ? slaveDataSourceConfigurations.get(0) : null;
    }

    public boolean hasSoleSlave() {
        return 1 == slaveDataSourceConfigurations.size();
    }

    protected void ensureEssentialDatasourceConfig(String dbId, Map<String, String> conf) {
        if (null == username) {
            LOGGER.warn("No data source user configuration specified. Will use the default 'sa' user");
            username = "sa";
            conf.put("username", username);
        }

        if (null == password) {
            password = "";
        }

        if (null == url) {
            try {
                Class.forName("org.h2.Driver");
                LOGGER.warn("No database URL configuration specified. Will use the default h2 inmemory test database");
                String dbName = "test" + (DbServiceManager.DEFAULT.equals(dbId) ? "" : "-" + dbId);
                url = "jdbc:h2:./" + dbName;
                conf.put("url", url);
            } catch (ClassNotFoundException e) {
                throw new ConfigurationException("No database URL configuration specified to db service: " + id);
            }
        }

        if (null == driver) {
            if (url.contains("mysql")) {
                try {
                    driver = "com.mysql.cj.jdbc.Driver";
                    Class.forName(driver);
                } catch (ClassNotFoundException e) {
                    // we are using mysql jdbc driver 5.x
                    driver = "com.mysql.jdbc.Driver";
                }
            } else if (url.contains("postgresql")) {
                driver = "org.postgresql.Driver";
            } else if (url.contains("jdbc:h2:")) {
                driver = "org.h2.Driver";
            } else if (url.contains("jdbc.mariadb")) {
                driver = "org.mariadb.jdbc.Driver";
            } else if (url.contains("jdbc:oracle")) {
                driver = "oracle.jdbc.OracleDriver";
            } else if (url.contains("sqlserver")) {
                driver = "com.microsoft.sqlserver.jdbc.SQLServerDriver";
            } else if (url.contains("jdbc:db2")) {
                driver = "com.ibm.db2.jcc.DB2Driver";
            } else {
                throw E.invalidConfiguration("JDBC driver needs to be configured for datasource: %s", id);
            }
            conf.put("driver", driver);
            LOGGER.warn("JDBC driver not configured, system automatically set to: " + driver);
        }
    }

    private void init(String dbId, Map<String, String> conf, boolean readOnly) {
        username = get(conf, SQL_CONF_USERNAME, username);
        password = get(conf, SQL_CONF_PASSWORD, password);
        driver = get(conf, SQL_CONF_DRIVER);
        url = get(conf, SQL_CONF_URL, url);

        minConnections = getInt(conf, SQL_CONF_MIN_CONNECTIONS, DEF_MIN_CONN);
        maxConnections = getInt(conf, SQL_CONF_MAX_CONNECTIONS, DEF_MAX_CONN);

        connectionTimeout = getInt(conf, SQL_CONF_CONNECTION_TIMEOUT, DEF_CONN_TIMEOUT);

        autoCommit = getBoolean(conf, SQL_CONF_AUTO_COMMIT, readOnly);
        this.readOnly = !readOnly ? getBoolean(conf, SQL_CONF_READONLY, false) : readOnly;

        String isoLevel = get(conf, "isolationLevel", String.valueOf(this.isolationLevel));
        if (null != isoLevel) {
            if (N.isInt(isoLevel)) {
                this.isolationLevel = Integer.parseInt(isoLevel);
            } else {
                isoLevel = Keyword.of(isoLevel.toUpperCase()).constantName();
                if (isoLevel.startsWith("TRANSACTION_")) {
                    isoLevel = isoLevel.substring("TRANSACTION_".length());
                }
                E.invalidConfigurationIf(!isolationLevels.containsKey(isoLevel), "Unknown isolationLevel: %s", isoLevel);
                this.isolationLevel = isolationLevels.get(isoLevel);
            }
        }
        ensureEssentialDatasourceConfig(dbId, conf);
        customProperties = conf;
    }

    private static final Map<String, Integer> isolationLevels = C.Map(
            "NONE", Connection.TRANSACTION_NONE,
            "READ_UNCOMMITTED", Connection.TRANSACTION_READ_UNCOMMITTED,
            "READ_COMMITTED", Connection.TRANSACTION_READ_COMMITTED,
            "REPEATABLE_READ", Connection.TRANSACTION_REPEATABLE_READ,
            "SERIALIZABLE", Connection.TRANSACTION_SERIALIZABLE
    );

    private static String get(Map<String, String> conf, String key, String def) {
        return conf.containsKey(key) ? conf.get(key) : def;
    }

    private static String get(Map<String, String> conf, String key) {
        return conf.get(key);
    }

    private static int getInt(Map<String, String> conf, String key, int def) {
        return conf.containsKey(key) ? parseInt(get(conf, key)) : def;
    }

    private static int parseInt(String s) {
        E.illegalArgumentIf(S.blank(s));
        List<String> parts;
        N.Op op;
        int seed;
        if (s.contains("*")) {
            parts = S.fastSplit(s, "*");
            op = N.Op.MUL;
            seed = 1;
        } else if (s.contains("+")) {
            parts = S.fastSplit(s, "+");
            op = N.Op.ADD;
            seed = 0;
        } else {
            return Integer.parseInt(s);
        }
        int i = seed;
        for (String part : parts) {
            i = op.apply(i, Integer.parseInt(part)).intValue();
        }
        return i;
    }

    private static boolean getBoolean(Map<String, String> conf, String key, boolean def) {
        return conf.containsKey(key) ? Boolean.parseBoolean(get(conf, key)) : def;
    }
}
